src/Oxpecker/ModelValidation.fs

```diff
@@ -0,0 +1,137 @@
+namespace Oxpecker
+
+[<AutoOpen>]
+module ModelValidation =
+
+    open System.Collections.Generic
+    open System.ComponentModel.DataAnnotations
+    open System.Runtime.CompilerServices
+    open Microsoft.AspNetCore.Http
+
+    type ValidationErrors(errors: ResizeArray<ValidationResult>) =
+        let errorDict =
+            lazy
+                (let dict = Dictionary<string, ResizeArray<string|null>>()
+                 for error in errors do
+                     for memberName in error.MemberNames do
+                         match dict.TryGetValue(memberName) with
+                         | true, value -> value.Add(error.ErrorMessage)
+                         | false, _ ->
+                             let arrayList = ResizeArray(1)
+                             arrayList.Add(error.ErrorMessage)
+                             dict[memberName] <- arrayList
+                 dict)
+        member this.All: ValidationResult seq = errors
+        member this.ErrorMessagesFor(name) : seq<string|null> =
+            match errorDict.Value.TryGetValue(name) with
+            | true, value -> value
+            | false, _ -> Seq.empty
+
+    type InvalidModel<'T> = 'T * ValidationErrors
+
+    [<RequireQualifiedAccess>]
+    type ModelState<'T> =
+        | Empty
+        | Valid of 'T
+        | Invalid of InvalidModel<'T>
+        member this.Value(f: 'T -> string | null) =
+            match this with
+            | Empty -> null
+            | Valid model -> f model
+            | Invalid(model, _) -> f model
+        member this.BoolValue(f: 'T -> bool) =
+            match this with
+            | Empty -> false
+            | Valid model -> f model
+            | Invalid(model, _) -> f model
```

Comment:
```suggestion
        private member this.Apply(f: 'T -> 'U) =
            match this with
            | Empty -> Unchecked.defaultof<'U>
            | Valid model -> f model
            | Invalid(model, _) -> f model
        member this.Value(f: 'T -> string | null) = this.Apply f
        member this.BoolValue(f: 'T -> bool) = this.Apply f
```

Reply:
While I get the duplication point, I like the readability of the current implementation more (and it doesn't use `Unchecked` module)